subprojects/core/src/test/groovy/org/gradle/api/internal/AbstractDomainObjectCollectionSpec.groovy

```diff
@@ -245,6 +244,27 @@ abstract class AbstractDomainObjectCollectionSpec<T> extends Specification {
         0 * _
     }
 
+    def "provider for iterable of elements is queried when .all configuration is added first"() {
+        containerAllowsExternalProviders()
+        def provider = Mock(CollectionProviderInternal)
+        def seen = []
+
+        given:
+        addToContainer(b)
+        container.all {
+            seen << it
+        }
+
+        when:
+        container.addAllLater(provider)
+
+        then:
+        _ * provider.getElementType() >> getType()
+        _ * provider.get() >> [a, d]
```

Comment:
I don't understand where the expectation was set on the mock `provider` to return `a` and `d`.  The test is verifying it here, so it must have been set somewhere.

Reply:
This is a Spock thing.  This is the expectation and verify statement.  Spock rewrites things to setup the mock earlier. 

`_ * provider.get()` means that this method can be called any number of times and `provider.get() >> [a, d]` means that it always returns `[a, d]`.

`0 * _` prevents any other mocks from being called.